cli/lib/kontena/cli/master/login_command.rb

```diff
@@ -239,44 +206,127 @@ def response_from_web_flow
+        server.name = next_name(response['server']['name'])
       else
-        nil
+        server.name = next_default_name
       end
     end
 
-    def update_server_token(server, response)
-      server.token = Kontena::Cli::Config::Token.new
-      server.token.access_token = response['access_token']
-      server.token.refresh_token = response['refresh_token']
-      server.token.expires_at = in_to_at(response['expires_in'])
-      server.token.username = response.fetch('user', {}).fetch('name', nil) || response.fetch('user', {}).fetch('email', nil)
+    def update_server_username(server, response)
+      return nil unless response.kind_of?(Hash)
+      return nil unless response['user']
+      server.token.username = response['user']['name'] || response['user']['email']
       server.username = server.token.username
     end
 
-    def update_server_name(server, response)
-      return unless server.name.nil?
-
-      if self.name
-        server.name = self.name
-      elsif response['server'] && response['server']['name']
-        server.name = response['server']['name']
-      elsif config.find_server('kontena-master')
-        new_name = "kontena-master-2"
-        new_name.succ! until config.find_server(new_name).nil?
-        server.name = new_name
+    def update_server_token(server, response)
+      if !response.kind_of?(Hash)
+        raise TypeError, "Response type mismatch - expected Hash, got #{response.class}"
+      elsif response['code']
+        use_authorization_code(server, response['code'])
+      elsif response['error']
+        exit_with_error "Authentication failed: #{response['error']} #{response['error_description']}"
       else
-        server.name = "kontena-master"
+        server.token = Kontena::Cli::Config::Token.new
+        server.token.access_token  = response['access_token']
+        server.token.refresh_token = response['refresh_token']
+        server.token.expires_at    = response['expires_at']
+      end
+    end
+
+    def update_server_to_config(server)
+      server.name ||= next_default_name
+      config.servers << server unless config.servers.include?(server)
+      config.current_server = server.name
+      config.write
+      config.reset_instance
+    end
+
+    # Figure out or create a server based on url or name.
+    #
+    # No name or url provided: try to use current_master
+    # A name provided with --name but no url defined: try to find a server by name from config
+    # An URL starting with 'http' provided: try to find a server by url from config
+    # An URL not starting with 'http' provided: try to find a server by name
+    # An URL and a name provided
+    #  - If a server is found by name: use entry and update URL to the provided url
+    #  - Else create a new entry with the url and name
+    #
+    # @param name [String] master name
+    # @param url [String] master url or name
+    # @return [Kontena::Cli::Config::Server]
+    def select_a_server(name, url)
```

Comment:
Lets implement dedicated methods to select server with name and url or only with name or url. Would make this method smaller (and maybe cleaner when hiding those dirty details).